x/ecocredit/migrations/v3/core.go

```diff
@@ -0,0 +1,385 @@
+	if err != nil {
+		return err
+	}
+
+	classInfoTable := classInfoTableBuilder.Build()
+	batchInfoTableBuilder, err := orm.NewPrimaryKeyTableBuilder(BatchInfoTablePrefix, storeKey, &BatchInfo{}, cdc)
+	if err != nil {
+		return err
+	}
+	batchInfoTable := batchInfoTableBuilder.Build()
+
+	creditTypeSeqTableBuilder, err := orm.NewPrimaryKeyTableBuilder(CreditTypeSeqTablePrefix, storeKey, &CreditTypeSeq{}, cdc)
+	if err != nil {
+		return err
+	}
+	creditTypeSeqTable := creditTypeSeqTableBuilder.Build()
+
+	// migrate credit classes to ORM v1
+	classItr, err := classInfoTable.PrefixScan(sdkCtx, nil, nil)
+	if err != nil {
+		return err
+	}
+	defer classItr.Close()
+
+	classIDsMap := make(map[uint64]string)   // map of a credit classID to className
+	projectIDsMap := make(map[string]uint64) // map of a credit className to projectID
+	ctx := sdk.WrapSDKContext(sdkCtx)
+	for {
+		var classInfo ClassInfo
+		if _, err := classItr.LoadNext(&classInfo); err != nil {
+			if orm.ErrIteratorDone.Is(err) {
+				break
+			}
+			return err
+		}
+
+		admin, err := sdk.AccAddressFromBech32(classInfo.Admin)
+		if err != nil {
+			return err
+		}
+		dest := api.ClassInfo{
+			Name:       classInfo.ClassId,
+			Admin:      admin,
+			Metadata:   string(classInfo.Metadata),
+			CreditType: classInfo.CreditType.Abbreviation,
+		}
+		classID, err := ss.ClassInfoTable().InsertReturningID(ctx, &dest)
+		if err != nil {
+			return err
+		}
+		classIDsMap[classID] = classInfo.ClassId
+
+		// migrate class issuers to ORM v1
+		for _, issuer := range classInfo.Issuers {
+			addr, err := sdk.AccAddressFromBech32(issuer)
+			if err != nil {
+				return err
+			}
+
+			if err := ss.ClassIssuerTable().Insert(ctx, &api.ClassIssuer{
+				ClassId: classID,
+				Issuer:  addr,
+			}); err != nil {
+				return err
+			}
+		}
+
+		// delete class info from old store
+		if err := classInfoTable.Delete(sdkCtx, &ClassInfo{ClassId: classInfo.ClassId}); err != nil {
+			return err
+		}
+	}
+
+	// migrate credit type sequence to ORM v1
+	cItr, err := creditTypeSeqTable.PrefixScan(sdkCtx, nil, nil)
+	if err != nil {
+		return err
+	}
+	defer cItr.Close()
+
+	for {
+		var ctype CreditTypeSeq
+		if _, err := cItr.LoadNext(&ctype); err != nil {
+			if orm.ErrIteratorDone.Is(err) {
+				break
+			}
+
+			return err
+		}
+		if err := ss.ClassSequenceTable().Save(ctx, &api.ClassSequence{
+			CreditType:  ctype.Abbreviation,
+			NextClassId: ctype.SeqNumber,
+		}); err != nil {
+			return err
+		}
+
+		// delete credit type sequence from old store
+		if err := creditTypeSeqTable.Delete(sdkCtx, &ctype); err != nil {
+			return err
+		}
+	}
+
+	// migrate projects
+	// TODO: update with actual data
+	projects := []*api.ProjectInfo{
+		{
+			Name:            "P01",
+			ClassId:         1,
+			Admin:           sdk.AccAddress("cosmos154hmhstk3gpkv2ndec8zkjkc5c3svutcqcswne"),
+			ProjectLocation: "AB-CDE FG1 345",
+			Metadata:        "metadata",
+		},
+	}
```

Comment:
Is it possible to iterate over credit batches rather than manually inserting project information?

- For `Name`, we could track the number of projects per credit class and automatically add the project name using the same format when creating projects (i.e. `C0101` is the first project in the `C01` credit class).
- For `ClassId` the batch provides the class id.
- For `Admin`, the batch provides the issuer.
- For `ProjectLocation`, the batch provides the project location.
- For `Metadata`, we could leave blank and maybe we could squeeze in another pull request adding the ability to update the metadata field #979

We would also need to skip adding a new project if all fields (except for the project name) match a project that has already been inserted. Does this sound possible?

Reply:
> Is it possible to iterate over credit batches rather than manually inserting project information?

Yes. We can do this.